cmd/release-controller/http_candidate.go

```diff
@@ -0,0 +1,336 @@
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	switch req.URL.Query().Get("format") {
+	default:
+		var candidate *ReleaseCandidate
+		if releaseCandidateList[releaseStreamName] != nil && len(releaseCandidateList[releaseStreamName].Items) > 0 {
+			candidate = releaseCandidateList[releaseStreamName].Items[0]
+			upgradeSuccess := make([]string, 0)
+			upgrades := c.graph.UpgradesTo(candidate.FromTag)
+			for _, u := range upgrades {
+				if u.Total > 0 {
+					if float64(100*u.Success)/float64(u.Total) > successPercent {
+						upgradeSuccess = append(upgradeSuccess, u.From)
+					}
+				}
+			}
+			sort.Strings(upgradeSuccess)
+			candidate.UpgradeFrom = upgradeSuccess
+		}
+		data, err := json.MarshalIndent(candidate.ReleasePromoteJobParameters, "", "  ")
+		if err != nil {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+		fmt.Fprintf(w, string(data))
+	}
+}
+
+// z-stream and timestamp
+var rePreviousReleaseName = regexp.MustCompile("(?P<STREAM>.*)-(?P<TIMESTAMP>[0-9]{4}-[0-9]{2}-[0-9]{2}-[0-9]{6})")
+
+const releaseTimestampFormat = "2006-01-02-150405"
+
+func (c *Controller) findReleaseCandidates(releaseStreams ...string) (map[string]*ReleaseCandidateList, error) {
+	releaseCandidates := make(map[string]*ReleaseCandidateList)
+	if len(releaseStreams) == 0 {
+		return releaseCandidates, nil
+	}
+
+	releaseStreamTagMap, ok := c.findReleaseByName(true, releaseStreams...)
+	if !ok || len(releaseStreamTagMap) == 0 {
+		return releaseCandidates, fmt.Errorf("unable to find all release streams: %s", strings.Join(releaseStreams, ","))
+	}
+
+	// next release version name for each release stream
+	next := make(map[string]*semver.Version)
+	// creation time of nightly/image that latest stable release in release stream was promoted from
+	latestPromotedTime := make(map[string]int64)
+	type stableRef struct {
+		from string
+		name string
+		time int64
+	}
+	stableReleases := make([]stableRef, 0)
+	stable, err := c.stableReleases(false)
+	if err != nil {
+		return releaseCandidates, err
+	}
+	for _, r := range stable.Releases {
+		for _, tag := range r.Release.Source.Spec.Tags {
+			if tag.Annotations[releaseAnnotationSource] != fmt.Sprintf("%s/%s", r.Release.Source.Namespace, r.Release.Source.Name) {
+				continue
+			}
+			if _, err := semverParseTolerant(tag.Name); err == nil {
+				t, _ := time.Parse(time.RFC3339, tag.Annotations[releaseAnnotationCreationTimestamp])
+				stableReleases = append(stableReleases, stableRef{from: tag.From.Name, name: tag.Name, time: t.Unix()})
+			}
+		}
+	}
+	sort.Slice(stableReleases, func(i, j int) bool {
+		vi, _ := semverParseTolerant(stableReleases[i].name)
+		vj, _ := semverParseTolerant(stableReleases[j].name)
+		return vi.GT(vj)
+	})
+	remaining := len(releaseStreams)
+	for _, r := range stableReleases {
+		if remaining == 0 {
+			break
+		}
+		v, _ := semverParseTolerant(r.name)
```

Comment:
You're ignoring errors here but r.name is either semver or not. If it's not, code below this shouldn't be trying to use v in its uninitialized state.

Where is `name` coming from, I don't see you defining it?

Reply:
https://github.com/openshift/release-controller/pull/103/files#diff-a17a0bcbe6c66848f22805e6bff079ddR233
It's part of the stableRef struct